tls: Introduce OpenSSL

michael-redpanda opened the pull request:
Introduces OpenSSL as an alternative TLS implementation to GnuTLS.  This is a build-time configuration controlled by the CMake variable `Seastar_USE_OPENSSL`.  The `configure.py` script has been updated to now have a `--crypto-provider` option.  Valid arguments to that are `OpenSSL` and `GnuTLS`.

This implementation was released in [Redpanda v24.2](https://github.com/redpanda-data/redpanda/releases/tag/v24.2.1) on July 31st, and has been running on production clusters since.

Redpanda implemented these changes in order to provide a FIPS-compliant build to customers that require it (such as those wishing to undergo FedRAMP evaluation).  OpenSSL was selected as it allows implementors to maintain the validation of the cryptographic module even when it's built from source.

* [OpenSSL's security policy](https://csrc.nist.gov/CSRC/media/projects/cryptographic-module-validation-program/documents/security-policies/140sp4282.pdf)
  * See appendix A for instructions on how to build OpenSSL for FIPS
* [FIPS 140-2 Implementation Guidance](https://csrc.nist.rip/csrc/media/projects/cryptographic-module-validation-program/documents/fips140-2/fips1402ig.pdf)
  * See section G.5 Maintaining validation compliance of software or firmware cryptographic
modules

No changes have been introduced to enable the FIPS provider for Seastar.  It is up to the implementor to enable and use the FIPS cryptographic module if desired.

avikivity commented:
@elcallio please review

Are there any functional differences?

Is hot reload of certificates supported?

Should we support gnutls and openssl in parallel?

michael-redpanda commented:
> Are there any functional differences?

The only major difference between the OpenSSL vs GnuTLS implementation is how the implementation is configured.  With OpenSSL, there are 5 new methods that can be used to control it (compared to the single `set_priority_string` with GnuTLS):

* `set_cipher_string` - used for controlling which ciphers are available in TLS1.2 and below ([OpenSSL SSL_CTX_set_cihper_list](https://docs.openssl.org/master/man3/SSL_CTX_set_cipher_list/))
* `set_ciphersuites` - used for controlling which cipher suites are available in TLS1.3 (same doc page as above)
* `enable_server_precedence`
* `set_minimum_tls_version` - defaults to 1.0
* `set_maximum_tls_version` - defaults to 1.3

> Is hot reload of certificates supported?

Yes

elcallio commented on src/net/tls-impl.hh:
```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright 2015 Cloudius Systems
+ */
+#pragma once
```
Why is this file in public header dir? Seems to me it should only be required by internal compilation units, so should probably be under src.

elcallio replied:
Doh, sorry, github confused me.

elcallio commented on src/net/ossl.cc:
```diff
@@ -815,11 +818,17 @@ class session : public enable_shared_from_this<session>, public session_impl {
     // This function waits for the _output_pending future to resolve
     // If an error occurs, it is saved off into _error and returned
     future<> wait_for_output() {
+        _logger.trace("wait_for_output");
```
Why invent a new logger type when you could just add a formatter for the session object that pretty-prints the local/remote etc, and then just use `tls_log::trace("{} wait_for_output")`?

michael-redpanda replied:
Thanks for the suggestion, great idea!

elcallio commented on tests/unit/tls_test.cc:
```diff
@@ -1039,8 +1039,12 @@ SEASTAR_THREAD_TEST_CASE(test_reload_certificates) {
+                // during the handshake before the client has fully
+                // closed its connection, then the get() call will
+                // succeed by return an empty buffer indicating EOF
+                BOOST_REQUIRE(res.size() == 0);
```
Seems like a good change, but not related to the rest of the PR

michael-redpanda replied:
This test was being flaky when using OpenSSL.  It would pass reliably in release but not in debug.

tchaikov replied:
nit, might want to use `BOOST_REQUIRE_EQUAL()` for better postmortem debugging experience. as Boost.Test prints out the lhs and rhs of the comparison f the check fails.

avikivity commented:
The different configuration options mean one of two:
 - side-by-side support, at least for a transitional period
 - emulate priority strings for openssl (far from the focus of Seastar)

So I think we should choose one implementation, and deprecate the other after a transition period.

gnutls was chosen due to funky licensing and an unstable API from openssl, but I think that's behind us now. Given that, it's better to use the leader in the field rather than a follower.

mykaul commented:
GnuTLS is not FIPS enabled, if compiled with it? (I see https://www.gnutls.org/manual/html_node/FIPS140_002d2-mode.html ) - what's missing?

michael-redpanda commented:
> GnuTLS is not FIPS enabled, if compiled with it? (I see https://www.gnutls.org/manual/html_node/FIPS140_002d2-mode.html ) - what's missing?

The FIPS flag for GnuTLS means that GnuTLS will work in a FIPS _compliant_ way (e.g. rejecting any non FIPS approved crypto like DES or GOST), however that doesn't mean that its implementation was _validated_.  GnuTLS doesn't provide a path to build it and maintain validation, OpenSSL does (see above links to OpenSSL's security policy).

elcallio commented:
> gnutls was chosen due to funky licensing and an unstable API from openssl, but I think that's behind us now. Given that, it's better to use the leader in the field rather than a follower.

While I can somewhat sympathize with the sentiment, this would effectively mean dropping the way we handle TLS config for clients/applications (i.e. usage of the prio string). Since at least one database application I know of exposes this in its customer config, you'd effectively be asking to require changing all customer TLS configs where such a prio is applied. Not sure how many they are and how complicated the configs are. You said a transition period, but not sure how to handle this, nor enforce a config migration with clients?

avikivity commented:
> > gnutls was chosen due to funky licensing and an unstable API from openssl, but I think that's behind us now. Given that, it's better to use the leader in the field rather than a follower.
> 
> While I can somewhat sympathize with the sentiment, this would effectively mean dropping the way we handle TLS config for clients/applications (i.e. usage of the prio string). Since at least one database application I know of exposes this in its customer config, you'd effectively be asking to require changing all customer TLS configs where such a prio is applied. Not sure how many they are and how complicated the configs are. You said a transition period, but not sure how to handle this, nor enforce a config migration with clients?

I would not want to force a config migration.

Claude says this:

> The gnutls-cli tool includes a "--priority-debug" option that can show detailed information about a priority string's settings, but there isn't a direct tool to convert between GnuTLS priority strings and OpenSSL configuration formats.
> 
> You can analyze the GnuTLS priority string components with `gnutls-cli --priority-debug "PRIORITY STRING"` and then manually map those to equivalent OpenSSL configurations based on the cipher suites and protocols shown.
> 
> For more control, you could write a script that parses the gnutls-cli output and maps the components to OpenSSL's configuration syntax, though you'd need to account for the differences in how each library names and groups their ciphers and protocols.

elcallio commented:
Yes, but I am honestly very nervous about writing/maintaining a prio string translator. The mapping is not just cipher to cipher etc, it is a state machine in itself, disabling and adding ciphers, exchange modes etc. 
As for using cli tool and manually map - that is what a customer would have to do if we changed the config approach.

avikivity commented:
> Yes, but I am honestly very nervous about writing/maintaining a prio string translator. The mapping is not just cipher to cipher etc, it is a state machine in itself, disabling and adding ciphers, exchange modes etc. As for using cli tool and manually map - that is what a customer would have to do if we changed the config approach.

I agree with that.

@tzach do you have any insight about priority string configuration across our fleet? Do we ever diverge from the default?

mykaul commented:
> > Yes, but I am honestly very nervous about writing/maintaining a prio string translator. The mapping is not just cipher to cipher etc, it is a state machine in itself, disabling and adding ciphers, exchange modes etc. As for using cli tool and manually map - that is what a customer would have to do if we changed the config approach.
> 
> I agree with that.
> 
> @tzach do you have any insight about priority string configuration across our fleet? Do we ever diverge from the default?

Since we did not disable TLSv1.1 by default (not sure why), there's a good chance users do it - https://enterprise.docs.scylladb.com/stable/operating-scylla/security/client-node-encryption.html#priority-string-and-tlsv1-2-1-3-support

michael-redpanda commented on src/seastar.cc:
```diff
@@ -335,6 +335,7 @@ module : private;
 #include <seastar/net/virtio.hh>
 
 #include "net/native-stack-impl.hh"
+#include "net/tls-impl.hh"
```
This is necessary for when compiling with modules enabled

michael-redpanda commented on src/net/tls-impl.hh:
```diff
@@ -0,0 +1,236 @@
+public:
+    static std::unique_ptr<connected_socket_impl> get(connected_socket s) {
+        return std::move(s._csi);
+    }
```
This was just moved from [`tls.cc`](https://github.com/scylladb/seastar/blob/2bba1204dbb018ff1551d08fc80a951af918a37d/src/net/tls.cc#L68-L80) to this file.

elcallio replied:
Yes. But that _was_ a cc. This is a header.

michael-redpanda commented on src/net/tls-impl.hh:
```diff
@@ -0,0 +1,236 @@
+        if (_session && _session.use_count() == 1) {
+            _session->close();
+        }
+    }
```
same [here](https://github.com/scylladb/seastar/blob/2bba1204dbb018ff1551d08fc80a951af918a37d/src/net/tls.cc#L1833-L1840)

elcallio replied:
Same as above. Code should not be in headers if it can be avoided (even internal, not very included ones)

michael-redpanda commented on src/net/tls-impl.cc:
```diff
@@ -229,9 +231,33 @@ void tls::credentials_builder::set_session_resume_mode(session_resume_mode m) {
+#endif
+
+#ifdef SEASTAR_USE_OPENSSL
+void tls::credentials_builder::set_cipher_string(const sstring& cipher_string) {
```
That's a good idea, I'll move the definitions to the correct source file

michael-redpanda commented on src/net/ossl.cc:
```diff
@@ -0,0 +1,2023 @@
+
+        // This call is required to lower SSL's security level to permit TLSv1.0 and TLSv1.1
+        // See https://www.openssl.org/docs/man3.0/man3/SSL_CTX_set_security_level.html
+        SSL_CTX_set_security_level(ssl_ctx.get(), 0);
```
Did a little bit of research on this function and what the level controls changes depending on the version of OpenSSL.

For OpenSSL 3.0, setting security level to 3 disables TLS1.0 and below, and setting it to 4 disables TLS1.1 and below ([ref](https://docs.openssl.org/3.0/man3/SSL_CTX_set_security_level/#default-callback-behaviour)).

Starting with OpenSSL 3.1, level 1 disables TLS1.1 and below ([ref](https://docs.openssl.org/3.1/man3/SSL_CTX_set_security_level/#default-callback-behaviour)).

michael-redpanda commented:
Force push [`fc616ee`](https://github.com/scylladb/seastar/compare/5c709c28d63f9ab9dab5b253e45c352d9d6b0398..fc616ee38bc09af4c56fabd953e4fde29e879145):

* Added compile definition `SEASTAR_USE_GNUTLS` for when GnuTLS is the cryptographic provder
* Updated `#ifndef SEASTAR_USE_OPENSSL` to instead be `#ifdef SEASTAR_USE_GNUTLS`
* Move implementation specific credential builder methods to the implementation specific compilation units
* Removed `tls_session_logger` and added a pretty-print function for the OpenSSL implementation of `seastar::tls::session`
* Dynamically setting minimum security level passed to `SSL_CTX_set_security_level`

michael-redpanda commented:
Force push [`1a53361`](https://github.com/scylladb/seastar/compare/fc616ee38bc09af4c56fabd953e4fde29e879145..1a533615610fac2ae8fa02b2b5265f8926427188):

* Added missing compile definition for `SEASTAR_USE_GNUTLS` when compiled to use modules
* Added missing header include for `fmt/ostream.hh` when compiled to use modules

elcallio commented on include/seastar/net/tls.hh:
```diff
@@ -312,8 +318,8 @@ namespace tls {
         void set_client_auth(client_auth);
-        void set_priority_string(const sstring&);
         void set_session_resume_mode(session_resume_mode);
+        void set_priority_string(const sstring&);
```
Seems like a gratuitous change?

elcallio commented on include/seastar/net/tls.hh:
```diff
@@ -235,6 +247,12 @@ namespace tls {
         friend class reloadable_credentials;
         shared_ptr<impl> _impl;
+
+        // The following methods are provided so classes that inherity from
```
Nit: inherit

elcallio commented on src/net/ossl.cc:
```diff
@@ -0,0 +1,2073 @@
+            auto dn = extract_dn_information();
+            if (dn) {
+                std::string_view stat_str_view{stat_str};
+                if (stat_str_view.ends_with(" ")) {
```
Nit: I assume this is an openssl quirk of some sort? Maybe a while instead (instead of assuming just one space), or maybe even a proper back and front whitespace strip?

elcallio commented on src/net/ossl.cc:
```diff
@@ -0,0 +1,2073 @@
+        }
+
+        auto & min_tls_version = _creds->minimum_tls_version();
+        auto & max_tls_version = _creds->maximum_tls_version();
```
Nit: space between type and ref.

elcallio commented on src/net/ossl.cc:
```diff
@@ -0,0 +1,2073 @@
+            }
+        }
+
+        auto get_min_level = [&ssl_ctx]() {
```
Nit: Should be "max level" really, as we should consider us setting the max level we can use given the tls version bounds.

elcallio commented on src/net/ossl.cc:
```diff
@@ -0,0 +1,2073 @@
+BIO_METHOD* get_method() {
+    static thread_local bio_method_ptr method_ptr = [] {
+        auto ptr = tls::create_bio_method();
+        if (!ptr) {
```
Seems redundant - all paths in create throws if failed...

elcallio commented on src/websocket/server.cc:
```diff
@@ -126,6 +130,30 @@ future<> connection::process() {
 static std::string sha1_base64(std::string_view source) {
     unsigned char hash[20];
+
+#ifdef SEASTAR_USE_OPENSSL
```
Could we instead make the function an external/internal-header-decl, and move the full impl into gnutls/openssl compilation units? This just adds to an already bad separation of concern/api exposure.

elcallio reviewed (commented):
Code looks ok, but again, I doubt I can really verify it just by looking. :-( 
One huge downside with the openssl code is that imho it seems to make the interface code even cludgier - perhaps partially an aspect of the code it needs to emulate is from a gnutls universe, but I would still argue that a lot of the gnutls interfaces are a bit nicer. 
Thus I would worry a little about maintenance here.

michael-redpanda commented:
Force push [`2f935ca`](https://github.com/scylladb/seastar/compare/1a533615610fac2ae8fa02b2b5265f8926427188..2f935cac22094152c1ed51daf6d3edcbea087fad):

* Addressed issues from comments
* Created GnuTLS & OpenSSL implementations for SHA1-Base64 used by websocket
* Fixed nits

tchaikov commented on CMakeLists.txt:
```diff
@@ -89,6 +89,16 @@ if (NOT Seastar_SCHEDULING_GROUPS_COUNT MATCHES "^[1-9][0-9]*")
+  OFF)
+
+if (Seastar_USE_OPENSSL)
+  set(Seastar_USE_GNUTLS OFF)
+else()
+  set(Seastar_USE_GNUTLS ON)
```
could you please add space after `set` and `else`? just to be consistent with the rest of this CMake script.

tchaikov commented on src/net/tls-impl.cc:
```diff
@@ -0,0 +1,740 @@
+ * under the License.
+ */
+/*
+ * Copyright 2015 Cloudius Systems
```
This header looks like a copy-and-paste. But in general, the copyright notice should reflect the original employer under whom this code was created, as work produced during employment typically belongs to the employer. If, and only *if*, there was a subsequent transfer of copyright to ScyllaDB (formerly known as Cloudius Systems), please update the header to use the current company name 'ScyllaDB' rather than 'Cloudius Systems'.

tchaikov commented on src/net/ossl.cc:
```diff
@@ -79,6 +79,8 @@ module seastar;
 #include "net/tls-impl.hh"
 #endif
 
+template <> struct fmt::formatter<seastar::tls::session> : fmt::ostream_formatter {};
```
is it possible to implement the specialization of `fmt::fornmatter<seastar::tls::session>` without the `operator<<`? it would be better if we can avoid adding more `operator<<`, for two reasons:

1. avoid adding the unused `operator<<`
2. avoid the overhead of using a temporary ostream for formatting `seastar::tls::session`.

tchaikov commented on src/net/ossl.cc:
```diff
@@ -0,0 +1,2068 @@
+}
+
+template<>
+struct fmt::formatter<seastar::ossl_errc> : public fmt::formatter<std::string_view> {
```
instead of inheriting from `fmt::formatter<std::string_view>` , i'd suggest inheriting from `fmt::formatter<string_view>`. please see https://github.com/scylladb/scylladb/commit/168ade72f8137b84b1f07b11d7d1bafd25adf5ba for more details. we switched from this approach to the proposed one in the referenced commit in scylladb a while ago.

tchaikov commented on src/net/ossl.cc:
```diff
@@ -770,6 +772,8 @@ class session : public enable_shared_from_this<session>, public session_impl {
     session(session_type t, shared_ptr<tls::certificate_credentials> creds,
             std::unique_ptr<net::connected_socket_impl> sock, tls_options options = {})
       : _sock(std::move(sock))
+      , _local_address(fmt::format("{}", _sock->local_address()))
```
nit, could use `fmt::to_string(_sock->local_address())` instead. simpler this way. and slightly more performant.

tchaikov commented on src/net/ossl.cc:
```diff
@@ -770,6 +772,8 @@ class session : public enable_shared_from_this<session>, public session_impl {
             std::unique_ptr<net::connected_socket_impl> sock, tls_options options = {})
       : _sock(std::move(sock))
+      , _local_address(fmt::format("{}", _sock->local_address()))
+      , _remote_address(fmt::format("{}", _sock->remote_address()))
```
ditto.

tchaikov commented on src/net/ossl.cc:
```diff
@@ -837,11 +841,25 @@ class session : public enable_shared_from_this<session>, public session_impl {
+                   session.get_type_string(),
+                   session.local_address(),
+                   session.remote_address());
+        return os;
```
as explained above, i'd prefer avoid adding more `operator<<` overloads unless they are actually necessary and used.

tchaikov commented on src/net/ossl.cc:
```diff
@@ -1084,17 +1118,21 @@ class session : public enable_shared_from_this<session>, public session_impl {
         return _in.get()
           .then([this](buf_type buf) {
               // Set EOF if it's empty
+              tls_log.debug("{} wait_for_input: buffer {}empty", *this, buf.empty() ? "is ": "");
```
i don't really understand this logging message:

- if `buf` is empty, we print "wait_for_input: buffer is empty" -- looks good
- if `buf` is not empty, we print "wait_for_input: buffer empty" -- what does this mean? shall we print "wait_for_input: buffer is not empty" or "wait_for_input: buffer not empty" instead?

michael-redpanda replied:
Thanks, that was my mess up.  Fixing it

tchaikov commented on .github/workflows/tests.yaml:
```diff
@@ -47,3 +49,16 @@ jobs:
+      mode: debug
+      enables: --enable-cxx-modules
+      enable-ccache: false
+      crypto_provider: OpenSSL
```
piggyback the tests with OpenSSL / GnuTLS might not be a great idea. because, we only enable the test step in https://github.com/scylladb/seastar/blob/613d8b39e0d6e8af4b4aff5cffb639a7016c8f7c/.github/workflows/test.yaml#L110-L112 if "cxx-modules" is *not* enabled. in other words, these two jobs do not run any of the unit tests. so i'd suggest adding a dedicated job in `tests.yaml` for building with the OpenSSL backend, and keep the existing job of `build_with_cxx_modules` intact. 

please allow me to provide more context here: because C++20 modules is currently an experimental feature, not all Seastar facilities are exposed as in the "seastar" C++20 module at this moment, and we only have a single "hello_cxx_module" test for testing the build with C++20 module. none of the unit tests is built with the "seastar" C++20 module at the time of writing.

tchaikov commented on src/net/ossl.cc:
```diff
@@ -0,0 +1,2068 @@
+            tls::error_category(),
+            msg};
+    } else {
+        auto err_code = static_cast<unsigned long>(error_codes.front());
+        if (ERR_LIB_SYS == ERR_GET_LIB(err_code)) {
+            // If the error code belongs to ERR_LIB_SYS, then the error is a system error
+            // Extract the errno using ERR_GET_REASON and throw a std::generic_category
+            return std::system_error(
+                ERR_GET_REASON(err_code),
+                std::generic_category(),
+                fmt::format("{}: {}", msg, error_codes));
+        }
+        return std::system_error(
+            static_cast<int>(err_code),
+            tls::error_category(),
+            fmt::format("{}: {}", msg, error_codes));
```
nit, could reduce the indent level by removing the "else" clause, because of the early return above.

tchaikov commented on src/net/ossl.cc:
```diff
@@ -0,0 +1,2068 @@
+
+    impl() : _creds([] {
+        auto store = X509_STORE_new();
+        if(store == nullptr) {
```
could you please add a space after `if`?

michael-redpanda commented:
Force push [db318da](https://github.com/scylladb/seastar/commit/db318daadfd31fadb14c45c8f128c1820a0a1a02):

* Addressed PR comments

michael-redpanda commented:
Force push [245d15c](https://github.com/scylladb/seastar/commit/245d15c8239a7db7f48c23e47dfe88171c716611):

* Rebased off master to fix merge conflicts introduced by https://github.com/scylladb/seastar/commit/6f39b89dbda812c7985918d3d8d6d7f7792b0619#diff-2f6778574276aef7b957daacf9904fc252fe1b44d45819bc89daafce9dac3972R1042

michael-redpanda commented:
Force push [cab469f](https://github.com/scylladb/seastar/commit/cab469f91be311fec34400502796a3eefe4c282c):

* Fixed typo

michael-redpanda commented:
Force push [47dfb11](https://github.com/scylladb/seastar/commit/47dfb11081910d5e70c02498d031633f66b1d1ec):

* Fixed logic of `get_security_level`
  * Returning anything >1 when using OpenSSL 3.0 will introduce key restrictions that may cause Seastar to be unable to connect with other endpoints
* Addressed error in `test_reload_certificates_with_only_shard0_notify`

avikivity commented:
Note test failures.

I'm conflicted here. On the one hand, the tls implementation has a user footprint in priority string or equivalent openssl config. On the other hand, that's a tiny difference and they're otherwise equivalent. OpenSSL would be my choice as the only supported library if there wasn't the user facing stuff.

Can we translate the priority string to openssl config? I asked Claude and it spewed a long python script, maybe that's good enough.

michael-redpanda commented:
> Note test failures

It appears all failures are from `Seastar.unit.rpc`, which I don't believe interacts with the TLS.  Were they flaky before?

> Can we translate the priority string to openssl config? I asked Claude and it spewed a long python script, maybe that's good enough.

It definitely could be done by setting all of these within an [OpenSSL config file](https://docs.openssl.org/master/man5/config/#openssl-library-configuration).  You'd just have to be careful to tell OpenSSL where to find it so it doesn't attempt to locate and use the system default one (unless that's desired).

michael-redpanda commented:
> Can we translate the priority string to openssl config? I asked Claude and it spewed a long python script, maybe that's good enough.

I guess your question is "can I take the input to`set_priority_string` and turn that into the appropriate OpenSSL API calls"?  It could be doable, but would take a while and I'm not sure if that's something anyone would really want to maintain.

avikivity commented:
> > Can we translate the priority string to openssl config? I asked Claude and it spewed a long python script, maybe that's good enough.
> 
> I guess your question is "can I take the input to`set_priority_string` and turn that into the appropriate OpenSSL API calls"? It could be doable, but would take a while and I'm not sure if that's something anyone would really want to maintain.

Is there no equivalent textual configuration?

I wouldn't want to maintain it, but neither do I want to maintain two different tls implementations.

michael-redpanda commented:
> Is there no equivalent textual configuration?

AFAIK only through the OpenSSL config file which can be loaded either automatically by OpenSSL when it is initialized or can be controlled by the user by calling [`OSSL_LIB_CTX_load_config`](https://docs.openssl.org/master/man3/OSSL_LIB_CTX/). And within there would be number of key-value pairs that would need to be set correctly.

avikivity commented:
I prefer replacing gnutls with openssl rather than adding yet another option. But perhaps we can't due to those minor configuration issues.

Maybe we can start a long deprecation cycle during which users will be encouraged to update their configuration.

avikivity commented on src/net/tls-impl.hh:
```diff
@@ -0,0 +1,232 @@
+    session_ref& operator=(const session_ref&) = default;
+
+    shared_ptr<session_impl> _session;
+};
```
`using session_ref = ` would be sufficient, no?

elcallio commented on src/net/tls-impl.hh:
```diff
@@ -0,0 +1,232 @@
+class tls_connected_socket_impl::source_impl: public data_source_impl, public session_ref {
+public:
+    using session_ref::session_ref;
+private:
```
socket_impl is stored with a unique ptr -> cannot be shared. Our IO design allows the input/output streams to exists independently of the creating socket. Thus the need to share the underlying session.

elcallio commented on src/net/tls-impl.hh:
```diff
@@ -0,0 +1,232 @@
+};
+
+class tls_connected_socket_impl : public net::connected_socket_impl, public session_ref {
+public:
```
Because it needs to create input/output sinks when asked.

avikivity commented on include/seastar/net/tls.hh:
```diff
@@ -212,13 +212,53 @@ namespace tls {
+         * for documentation on the format of the ciphersuites string.
+         */
+        void set_ciphersuites(const sstring&);
+
```
std::string_view is more flexible.

avikivity commented on include/seastar/net/tls.hh:
```diff
@@ -629,6 +685,12 @@ namespace tls {
     extern const int ERROR_NO_CIPHER_SUITES;
     extern const int ERROR_DECRYPTION_FAILED;
     extern const int ERROR_MAC_VERIFY_FAILED;
```
Are those error numbers compatible? And specified somewhere?

avikivity commented on include/seastar/net/tls.hh:
```diff
@@ -377,6 +428,11 @@ namespace tls {
+        sstring _ciphersuites;
+        bool _enable_server_precedence = false;
+        std::optional<tls_version> _min_tls_version;
+        std::optional<tls_version> _max_tls_version;
```
Let's package those in openssl/gnutls sub-structs.

michael-redpanda commented:
> If we indeed start a deprecation cycle, it will have to be run-time selectable (so users can update their configuration, change the selector, restart). I don't know how difficult that is.

I believe this would require a significant change to the PR to support this.  If this is what is desired, more than happy to look into it, but I would have to close this PR and open a new one.

michael-redpanda commented:
@avikivity  I haven't looked too deep into it yet, but I think making it runtime selectable would be a change to the API

avikivity commented:
> @avikivity I haven't looked too deep into it yet, but I think making it runtime selectable would be a change to the API

Why? If the API suits both openssl and gnutl statically, it should suit them dynamically, no?

michael-redpanda commented:
> > @avikivity I haven't looked too deep into it yet, but I think making it runtime selectable would be a change to the API
> 
> Why? If the API suits both openssl and gnutl statically, it should suit them dynamically, no?

The current implementation does not support building both at the same time, it has to be one or the other.  Both GnuTLS and OpenSSL implement many of the same named methods, e.g. `tls::wrap_client`, `tls::wrap_server`, etc.  Both also define the error codes like `tls::ERROR_UNKNOWN_COMPRESSION_ALGORITHM`, etc.

avikivity commented:
> > > @avikivity I haven't looked too deep into it yet, but I think making it runtime selectable would be a change to the API
> > 
> > 
> > Why? If the API suits both openssl and gnutl statically, it should suit them dynamically, no?
> 
> The current implementation does not support building both at the same time, it has to be one or the other. Both GnuTLS and OpenSSL implement many of the same named methods, e.g. `tls::wrap_client`, `tls::wrap_server`, etc.

Well, the entry points need to be disambiguated:

```c++
future<connected_socket> tls::wrap_client(shared_ptr<certificate_credentials> cred, connected_socket&& s, tls_options options) {
    session::session_ref sess(make_lw_shared<session>(session::type::CLIENT, std::move(cred), std::move(s),  options));
    connected_socket sock(std::make_unique<tls_connected_socket_impl>(std::move(sess)));
    return make_ready_future<connected_socket>(std::move(sock));
}
```

Here call a factory to create sess instead of hardcoding the implementation.


> Both also define the error codes like `tls::ERROR_UNKNOWN_COMPRESSION_ALGORITHM`, etc.


We'd change them from const globals to non-const globals intialized at library load time.

I don't want to cause unnecessary churn, but I also don't want to break running configurations.

avikivity commented:
> > > > @avikivity I haven't looked too deep into it yet, but I think making it runtime selectable would be a change to the API
> > > 
> > > 
> > > Why? If the API suits both openssl and gnutl statically, it should suit them dynamically, no?
> > 
> > 
> > The current implementation does not support building both at the same time, it has to be one or the other. Both GnuTLS and OpenSSL implement many of the same named methods, e.g. `tls::wrap_client`, `tls::wrap_server`, etc.
> 
> Well, the entry points need to be disambiguated:
> 
> ```c++
> future<connected_socket> tls::wrap_client(shared_ptr<certificate_credentials> cred, connected_socket&& s, tls_options options) {
>     session::session_ref sess(make_lw_shared<session>(session::type::CLIENT, std::move(cred), std::move(s),  options));
>     connected_socket sock(std::make_unique<tls_connected_socket_impl>(std::move(sess)));
>     return make_ready_future<connected_socket>(std::move(sock));
> }
> ```
> 
> Here call a factory to create sess instead of hardcoding the implementation.
> 
> > Both also define the error codes like `tls::ERROR_UNKNOWN_COMPRESSION_ALGORITHM`, etc.
> 
> We'd change them from const globals to non-const globals intialized at library load time.
> 
> I don't want to cause unnecessary churn, but I also don't want to break running configurations.

Another question is how to roll out a change across the cluster. I guess if the configuration is local to a node, you'd update it as part of each individual node upgrade.

michael-redpanda commented:
@avikivity If this is the direction your team wants to go, I may have to take this PR down and start from scratch.  I'll probably break this up into two separate PRs.  A 'staging' PR that will refactor the GnuTLS stuff out of `tls.hh` and then a second one to introduce OpenSSL.  Probably should have done that from the start.

avikivity commented:
> @avikivity If this is the direction your team wants to go, I may have to take this PR down and start from scratch. I'll probably break this up into two separate PRs. A 'staging' PR that will refactor the GnuTLS stuff out of `tls.hh` and then a second one to introduce OpenSSL. Probably should have done that from the start.

What were your plans wrt the small incompatibilities between openssl and gnutls? Do they not impact you?
